pkg/storage/segments/store.go

```diff
@@ -46,16 +46,95 @@ type Store interface {
+	ec            ecclient.Client
+	pdb           pointerdb.Client
+	rs            eestream.RedundancyStrategy
+	thresholdSize int
 }
 
 // NewSegmentStore creates a new instance of segmentStore
 func NewSegmentStore(oc overlay.Client, ec ecclient.Client,
-	pdb pointerdb.Client, rs eestream.RedundancyStrategy) Store {
-	return &segmentStore{oc: oc, ec: ec, pdb: pdb, rs: rs}
+	pdb pointerdb.Client, rs eestream.RedundancyStrategy, t int) Store {
+	return &segmentStore{oc: oc, ec: ec, pdb: pdb, rs: rs, thresholdSize: t}
+}
+
+// PeekThresholdReader allows a check to see if the size of a given reader
+// exceeds the maximum inline segment size or not.
+type PeekThresholdReader struct {
+	r              io.Reader
+	n              int // number of bytes read into thresholdBuf
+	thresholdBuf   []byte
+	totalReadBytes int
+}
+
+// NewPeekThresholdReader creates a new instance of PeekThresholdReader
+func NewPeekThresholdReader(r io.Reader) (pt *PeekThresholdReader) {
+	return &PeekThresholdReader{r: r, n: 0, thresholdBuf: nil, totalReadBytes: 0}
+}
+
+// Read initially reads bytes from the internal buffer, then continues
+// reading from the wrapped data reader. The number of bytes read `n`
+// is returned.
+func (pt *PeekThresholdReader) Read(p []byte) (n int, err error) {
+
+	// Case 1: if the total number of bytes read is greater than the number
+	// of bytes read into the thresholdBuf, then Read is called on the given
+	// byte slice.
+	if pt.totalReadBytes > pt.n {
+		return pt.r.Read(p)
+	}
+
+	thresholdBytesRemaining := pt.n - pt.totalReadBytes
+
+	// Case 2: if the length of the given byte slice p is less than or equal
+	// to the number of threshold bytes remaining to be read, then the slice
+	// of the threshold buffer from the end of totalReadBytes to the length
+	// of p is copied to p.
+	if len(p) <= thresholdBytesRemaining {
+		tmp := pt.thresholdBuf[pt.totalReadBytes : pt.totalReadBytes+len(p)]
+		copy(p, tmp)
+		pt.totalReadBytes += len(p)
+		return len(p), nil
+	}
+
+	// Case 3: The buffer tail slice is created then read.
+	// A slice of read bytes is copied to the given byte slice p.
+	tmp := pt.thresholdBuf[pt.totalReadBytes : pt.totalReadBytes+thresholdBytesRemaining]
+	bufTail := make([]byte, len(p)-thresholdBytesRemaining)
+	numTailBytes, err := pt.r.Read(bufTail)
+	if err != nil {
+		return 0, err
+	}
+	tmp = append(tmp, bufTail...)
+	copy(p, tmp)
+	n = thresholdBytesRemaining + numTailBytes
+	pt.totalReadBytes += n
+	return n, nil
+}
+
+// checkSize returns a bool to determine whether a reader's size
+// is inline-sized or not.
+func (pt *PeekThresholdReader) isInline(thresholdSize int) (inline bool, err error) {
+	err = pt.makeThresholdBuffer(thresholdSize)
+	if err != nil {
+		return false, err
+	}
+	if pt.n < thresholdSize {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (pt *PeekThresholdReader) makeThresholdBuffer(thresholdSize int) (err error) {
+	buf := make([]byte, thresholdSize)
+	n, err := pt.r.Read(buf)
```

Comment:
`Read` does not guarantee to make an effort to fill the buffer. It can easily read less than `len(buf)` bytes even if the reader is significantly larger.

Instead we should use `io.ReadFull`. Make sure to check carefully the docs: https://golang.org/pkg/io/#ReadFull